### PR TITLE
Get rid of 0 byte file if task get cancelled

### PR DESCRIPTION
**Pull request by jja725**
### What changes are proposed in this pull request?

always delete the file when cancelling the current file-output-stream

### Why are the changes needed?

Currently, the migration job that’s used by distributed-copy command doesn’t delete files when the output stream is cancelled due to an error. (Ref [here](https://github.com/Alluxio/alluxio/blob/7be4c6f6885a5092a3fb8621f27b5cc7d480d079/job/server/src/main/java/alluxio/job/plan/migrate/MigrateDefinition.java#L177)).

Stream cancellation doesn’t delete the file but simply suppresses completion of the file. (Ref [here](https://github.com/Alluxio/alluxio/blob/7be4c6f6885a5092a3fb8621f27b5cc7d480d079/core/client/fs/src/main/java/alluxio/client/file/AlluxioFileOutStream.java#L198)) So this means, distributed copy jobs will leave 0-byte files around when cancelled due to any error.

The partial fix is to always delete the file when cancelling the current file-output-stream within migration job.

### Does this PR introduce any user facing changes?
less 0 byte files

### CODE REVIEW

**Comment by yuzhu** [comment]
@jja725 how did you test this? could you add an integration test for cancellation?

File: tests/src/test/java/alluxio/client/cli/fs/command/DistributedCpCancelTest.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+  public void testDistributedCpCancelStats() throws Exception {
+
+    long jobId = sJobMaster.run(new MigrateConfig(
+        "/mnt/dir/file", "/testFileDest", "THROUGH", false));
+    // wait for execution until blocked by FileInStream in = fileSystem.openFile() in migration job
+    Thread.sleep(Constants.SECOND_MS * 3);
+    sJobShell.run("cancel", Long.toString(jobId));
+    JobTestUtils
+        .waitForJobStatus(sJobMaster, jobId, Sets.newHashSet(Status.CANCELED), TEST_TIMEOUT);
+    Assert.assertFalse(sFileSystem.exists(new AlluxioURI("/testFileDest")));
```
Comment (jja725) [inline tests/src/test/java/alluxio/client/cli/fs/command/DistributedCpCancelTest.java:71]: @yuzhu need an extra eye on this unit test. The delete operation also gets interrupted.

**Comment by jja725** [comment]
Error: 5.847 [ERROR] Errors: 
Error: 5.848 [ERROR] alluxio.server.ft.journal.raft.EmbeddedJournalIntegrationTestTransferLeadership.repeatedTransferLeadership
Error: 5.849 [ERROR]   Run 1: EmbeddedJournalIntegrationTestTransferLeadership.repeatedTransferLeadership:75->transferAndWait:250->EmbeddedJournalIntegrationTestBase.waitForQuorumPropertySize:58 » Timeout
Error: 5.849 [ERROR]   Run 2: EmbeddedJournalIntegrationTestTransferLeadership.repeatedTransferLeadership:75->transferAndWait:250->EmbeddedJournalIntegrationTestBase.waitForQuorumPropertySize:58 » Timeout

File: job/server/src/main/java/alluxio/job/plan/migrate/MigrateDefinition.java [hunk 2]
```diff
@@ -169,12 +169,15 @@ private static void migrate(MigrateCommand command,
...
       try (FileInStream in = fileSystem.openFile(new AlluxioURI(source), openFileOptions);
-           FileOutStream out = fileSystem.createFile(destinationURI, createOptions)) {
+          FileOutStream out = fileSystem.createFile(destinationURI, createOptions)) {
         try {
           IOUtils.copyLarge(in, out, new byte[8 * Constants.MB]);
         } catch (Throwable t) {
           try {
             out.cancel();
+            // this is to avoid mysterious interruption behavior on filesystem.delete
+            Thread.interrupted();
```
Comment (ggezer) [inline job/server/src/main/java/alluxio/job/plan/migrate/MigrateDefinition.java:179]: This is quite unusual. Can we uncover the mystery and avoid this hopefully?

Comment (jiacheliu3) [reply job/server/src/main/java/alluxio/job/plan/migrate/MigrateDefinition.java:179]: +1

Comment (jja725) [reply job/server/src/main/java/alluxio/job/plan/migrate/MigrateDefinition.java]: https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html
@ggezer @jiacheliu3 I think the problem here is we are throwing CancelException with wrapped Interrupted Exception so I have to check that flag according to the doc, update the code with corresponding change from

**Review by yuzhu — APPROVED** [review]
LGTM

**Comment by yuzhu** [comment]
alluxio-bot, merge this please

**Comment by jiacheliu3** [comment]
alluxio-bot, cherry-pick this to branch-2.7 please

**Pull request merged**